engine/overworld/weather.asm

```diff
@@ -0,0 +1,778 @@
+DoOverworldWeather:
+	push de
+	push hl
+	push bc
+	ld hl, wWeatherFlags
+	bit OW_WEATHER_DISABLED_F, [hl]
+	jr nz, .done
+	ld a, [wOverworldWeatherDelay]
+	and %1 ; 30 fps
+	jr z, .done
+	ld a, [wOverworldWeatherCooldown]
+	and a
+	jr z, .no_cooldown
+	dec a
+	ld [wOverworldWeatherCooldown], a
+.no_cooldown
+	ldh a, [hUsedOAMIndex]
+	cpl
+	add (SPRITEOAMSTRUCT_LENGTH * NUM_SPRITE_OAM_STRUCTS) + 1
+	ld hl, hUsedWeatherSpriteIndex
+	cp [hl]
+	jr nc, .ok
+	add -SPRITEOAMSTRUCT_LENGTH
+	ldh [hUsedWeatherSpriteIndex], a
+.ok
+	ld a, [wOverworldWeatherCooldown]
+	and a
+	jr nz, .on_cooldown
+	ld a, [wCurrentWeather]
+	dec a ; OW_WEATHER_RAIN
+	call z, DoOverworldRain
+	ld a, [wCurrentWeather]
+	cp OW_WEATHER_SNOW
+	call z, DoOverworldSnow
+	ld a, [wCurrentWeather]
+	cp OW_WEATHER_THUNDERSTORM
+	call z, DoOverworldRain
+	ld a, [wCurrentWeather]
+	cp OW_WEATHER_SANDSTORM
+	call z, DoOverworldSandstorm
+.done
+	ld hl, wOverworldWeatherDelay
+	inc [hl]
+	call WeatherSpriteLimitCheck
+	pop bc
+	pop hl
+	pop de
+	ret
+
+.on_cooldown
+	ld a, [wPrevWeather]
+	cp OW_WEATHER_RAIN
+	jr z, .rain_cooldown
+	cp OW_WEATHER_THUNDERSTORM
+	jr z, .rain_cooldown
+	cp OW_WEATHER_SNOW
+	jr z, .snow_cooldown
+	call DoSandFall
+	jr .cooldown_cleanup
+.snow_cooldown
+	call DoSnowFall
+	jr .cooldown_cleanup
+.rain_cooldown
+	call DoRainFall
+	call RainSplashCleanup
+.cooldown_cleanup
+	ld a, [wOverworldWeatherCooldown]
+	dec a
+	jr nz, .done
+	call ClearWeather
+	call LoadWeatherGraphics
+	jr .done
+
+SpawnRandomWeatherFullScreen::
+	lb bc, SCREEN_WIDTH_PX, SCREEN_HEIGHT_PX
+; fallthrough
+SpawnRandomWeatherCoords::
+; This is used to randomize the weather's starting position within an area
+; assumes minimum sprite x/y coords are 0.
+; bc = max screen x/y coords
+	ld a, [wCurrentWeather]
+	and a
+	ret z
+	ld a, [wCurrentWeather]
+	dec a ; rain
+	jr z, .rain
+.snow_loop
+	push bc
+	call ScanForEmptyOAM
+	pop bc
+	ret c
+	ld a, c
+	call RandomRange
+	ld [hli], a
+	ld a, b
+	call RandomRange
+	ld [hli], a
+	ld a, SNOWFLAKE_TILE
+	ld [hli], a
+	ld a, 6 ; pallete 6
```

Comment:
"palette"; although since multiple places in the code hard-code pal 6 for weather, I think this should be a constant (`PAL_OW_WEATHER`?).
```suggestion
	ld a, 6 ; pallete 6
```